### ISSUE TITLE
Add-Content and Out-File -Append can fail quietly, seemingly connected to cross-thread concurrency

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#14416]
Note:

* I don't know what other cmdlets are affected. _Update_: at least `Out-File -Append` is affected too.

* A common use case where the problem can surface is in the context of `ForEach-Object -Parallel` (while the use of `Add-Content` in parallel threads requires explicit synchronization, neglecting to do so should at least _report_ the  failure to write.)

## Steps to reproduce

```powershell
$tempFile = New-TemporaryFile
Write-Verbose -vb "Waiting for an Add-Content call to fail..."
$Error.Clear()
while ($true) {
  Remove-Item -ea Ignore $tempFile
  # Let multiple threads try to update Add-Content simultaneously.
  1..100 | ForEach-Object -Parallel { Add-Content $using:tempFile -Value $_ -ErrorAction Stop; if (-not $?) { Write-Warning $_.ToString() } }
  # On occasion, the file ends up with fewer than 100 lines.
  $countWritten = (Get-Content $tempFile).Count
  if ($countWritten -ne 100) {
    Write-Verbose -vb "$countWritten lines were written."
    # If fewer lines were written, an error should have been reported.
    $Error.Count | Should -BeGreaterOrEqual 1
  }
}
Remove-Item $tempFile
```

## Expected behavior

The tests should run indefinitely, because the `Add-Content` calls should either all succeed or those that fail should report an error.

## Actual behavior

A test eventually (may take a while) fails, because an `Add-Content` calls fails _silently_; e.g.:

```none
VERBOSE: 99 lines were written.
InvalidResult:
Line |
  13 | $Error.Count | Should -BeGreaterOrEqual 1
     | Expected the actual value to be greater than or equal to 1, but got 0.
```

## Environment data

```none
PowerShell Core 7.2.0-preview.1
```

**Comment by iSazonov** [comment]
@mklement0 Can you catch errors in `ForEach-Object -Parallel` scriptblock? Maybe add Get-Error?

**Comment by iSazonov** [comment]
I think really Get-Content does not fail. It write all values in the file but we have a race condition sometimes - some values are overwritten by subsequent ones.

**Comment by mklement0** [comment]
It's definitely a race condition.

_Some_ failure of `Add-Content` is involved and two possible reasons come to mind:

* The write operation fails, but the failure isn't reported.
* File content is _cached_ (which seems unlikely) and on writing there's a failure to recognize that the file has been modified by another thread.

Either way, a fix is required.

**Comment by mklement0** [comment]
Also note that the problem may not be cmdlet-specific, because `Out-File -Append` is affected too.

**Comment by iSazonov** [comment]
I doubt we can fix this since too many things in PowerShell are not thread-safe.
Perhaps the best option is to write this in the documentation if it is not already there.

@PaulHigin Could you please share your thoughts?

**Comment by PaulHigin** [comment]
Yes, PowerShell cmdlets are not and were never intended to be multi-threaded safe.  Users must beware of using cmdlets in multiple threads, whether via foreach -parallel or PowerShell API.  Same as with any non-multi-threaded-safe .NET object.

**Comment by mklement0** [comment]
Agreed in general, @PaulHigin, but while race conditions on _in-memory_ data (e.g., a non-concurrent collection) cannot be prevented, we are talking about concurrent _file access_ here.

And it seems to me that failure to update a file being written to by someone else should be _reported as such_ - the current, quiet failure is insidious. (Note that this is different from expecting the concurrency to be _managed_ automatically - that is indeed not the goal here.)

**Comment by SeeminglyScience** [comment]
Are you sure it's actually failing to write? It could be successfully writing input of similar length to the same offset in the file.

**Comment by iSazonov** [comment]
Safe solution was to mark cmdlets which is a thread-safe and allow in `ForEach-Object -Parallel` only such cmdlets. But in any case we can not make the `ForEach-Object -Parallel` fully thread-safe.

As for Add-Cmdlet, it has thick wrapper code and I guess it's not so easy to do it thread-safe.

**Comment by mklement0** [comment]
That's possible, @SeeminglyScience, but that is still a failure to synchronize access to a file for writing, which is something that should be addressed in general, irrespective of cross-_thread_ use, as it also applies cross-_process_.

Therefore, @iSazonov, this isn't about _thread safety_ per se. It's about not allowing any other entity - whether thread or process - write access to a file while it is being written to.

**Comment by ShivnarenSrinivasan** [comment]
*Not sure if I should create a separate issue for this--seemed similar enough to the main issue.*

# Redirection/Writing to file fails in `ForEach` block within function

_I'm a relative beginner, so might be a basic error--that said, **this same function worked earlier, and worked on*_ **`WindowsPowerShell`** (5.1)

```
function makeDirective([string]$directiveFile, $files) {

        if (Test-Path $directiveFile) {
            throw "File $directiveFile Exists"
        }
        ForEach ($file in $files) {
            "file $($file.Name)" >> $directiveFile
        }
}

# Files is obviously not empty
$files = Get-ChildItem <file_pattern>

makeDirective 'manifest.txt' $files
```

No errors, no warnings--but no file is outputted.
The same redirection line in the function works when I type it into a shell, or as a line in a script.

I'm using Version 7.1.4

**Comment by vexx32** [comment]
That would likely be a completely separate issue. FWIW I tried your function on 7.2.0-preview9 and it works as expected. I would guess something else is going on in your case, though I don't know what exactly.

**Comment by iSazonov** [comment]
Now .Net 6.0 has get new thread-safe file IO API.
https://devblogs.microsoft.com/dotnet/file-io-improvements-in-dotnet-6/#thread-safe-file-io

but I doubt that we will be able to use this.
